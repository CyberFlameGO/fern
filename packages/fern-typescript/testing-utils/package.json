{
  "name": "@fern-typescript/testing-utils",
  "version": "0.0.0",
  "repository": {
    "type": "git",
    "url": "git@github.com:fern-api/fern.git",
    "directory": "packages/fern-typescript/testing-utils"
  },
  "files": [
    "lib"
  ],
  "type": "module",
  "source": "src/index.ts",
  "module": "src/index.ts",
  "main": "src/index.ts",
  "sideEffects": false,
  "scripts": {
    "clean": "rm -rf ./lib && tsc --build --clean",
    "compile": "tsc --build",
    "test": "yarn run compile && jest --passWithNoTests",
    "lint:eslint": "eslint --max-warnings 0 . --ignore-path=../../../.eslintignore",
    "lint:eslint:fix": "eslint --max-warnings 0 . --ignore-path=../../../.eslintignore --fix",
    "format": "prettier --write --ignore-unknown --ignore-path ../../../shared/.prettierignore \"**\"",
    "format:check": "prettier --check --ignore-unknown --ignore-path ../../../shared/.prettierignore \"**\"",
    "depcheck": "depcheck"
  },
  "dependencies": {
    "@fern-api/ir-generator": "workspace:*",
    "@fern-api/workspace-loader": "workspace:*",
    "@fern-fern/ir-model": "0.0.49",
    "@fern-typescript/commons": "workspace:*",
    "execa": "^5.1.1",
    "is-ci": "^3.0.1",
    "memfs": "^3.4.1",
    "typescript": "4.6.4"
  },
  "devDependencies": {
    "@babel/core": "^7.18.9",
    "@babel/preset-env": "^7.18.9",
    "@babel/preset-typescript": "^7.18.6",
    "@types/is-ci": "^3.0.0",
    "@types/jest": "^28.1.6",
    "@types/node": "^18.6.1",
    "depcheck": "^1.4.3",
    "eslint": "^8.20.0",
    "jest": "^28.1.3",
    "prettier": "^2.7.1",
    "ts-jest": "^28.0.7"
  }
}
