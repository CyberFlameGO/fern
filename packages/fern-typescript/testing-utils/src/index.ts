export {
    installAndCompileGeneratedProject,
    installAndCompileGeneratedProjects,
} from "./installAndCompileGeneratedProject";
export { itFernETE } from "./itFernETE";
