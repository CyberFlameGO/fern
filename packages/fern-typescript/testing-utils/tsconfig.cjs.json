{
  "extends": "../../../shared/tsconfig.shared.json",
  "compilerOptions": { "composite": true, "outDir": "lib/cjs", "rootDir": "src", "module": "CommonJS" },
  "include": ["./src"],
  "references": [
    { "path": "../../cli/tsconfig.cjs.json" },
    { "path": "../../ir/ir-generator/tsconfig.cjs.json" },
    { "path": "../../workspace-parser/tsconfig.cjs.json" },
    { "path": "../commons/tsconfig.cjs.json" }
  ]
}
