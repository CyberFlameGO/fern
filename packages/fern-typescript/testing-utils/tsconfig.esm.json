{
  "extends": "../../../shared/tsconfig.shared.json",
  "compilerOptions": { "composite": true, "outDir": "lib/esm", "rootDir": "src", "module": "esnext" },
  "include": ["./src"],
  "references": [
    { "path": "../../cli/tsconfig.esm.json" },
    { "path": "../../compiler/compiler/tsconfig.esm.json" },
    { "path": "../commons/tsconfig.esm.json" }
  ]
}
