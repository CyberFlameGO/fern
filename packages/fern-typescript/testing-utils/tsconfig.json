{
  "extends": "../../../shared/tsconfig.shared.json",
  "compilerOptions": { "composite": true, "outDir": "lib", "rootDir": "src" },
  "include": ["./src"],
  "references": [
    { "path": "../../generation/ir/ir-generator" },
    { "path": "../../workspace-loader" },
    { "path": "../commons" }
  ]
}
