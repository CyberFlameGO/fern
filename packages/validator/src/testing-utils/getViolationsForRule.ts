import { parseWorkspaceDefinition } from "@fern-api/workspace-parser";
import { visitFernYamlAst } from "@fern-api/yaml-schema";
import { createAstVisitorForRules } from "../createAstVisitorForRules";
import { Rule, RuleViolation } from "../Rule";

export declare namespace getViolationsForRule {
    export interface Args {
        rule: Rule;
        absolutePathToDefinition: string;
    }
}

export async function getViolationsForRule({
    rule,
    absolutePathToDefinition,
}: getViolationsForRule.Args): Promise<RuleViolation[]> {
    const parseResult = await parseWorkspaceDefinition({
        name: undefined,
        absolutePathToDefinition,
    });
    if (!parseResult.didSucceed) {
        throw new Error("Failed to parse workspace");
    }

    const ruleRunner = await rule.create({ workspace: parseResult.workspace });
    const violations: RuleViolation[] = [];

    for (const [relativeFilePath, contents] of Object.entries(parseResult.workspace.files)) {
        const visitor = createAstVisitorForRules({
            relativeFilePath,
            contents,
            ruleRunners: [ruleRunner],
            addViolations: (newViolations) => {
                violations.push(...newViolations);
            },
        });
        await visitFernYamlAst(contents, visitor);
    }

    return violations;
}
