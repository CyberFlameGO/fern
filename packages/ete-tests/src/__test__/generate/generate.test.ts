import { installAndCompileGeneratedProject } from "@fern-typescript/testing-utils";
import execa from "execa";
import { readFile, rm } from "fs/promises";
import path from "path";

const FIXTURES_DIR = path.join(__dirname, "fixtures");

describe("fern generate tests", () => {
    itFixture("simple-model");
    // itFixture("hathora");
});

function itFixture(fixtureName: string) {
    it(
        // eslint-disable-next-line jest/valid-title
        fixtureName,
        async () => {
            const fixturePath = path.join(FIXTURES_DIR, fixtureName);
            const outputPath = path.join(fixturePath, "generated");
            await rm(outputPath, { force: true, recursive: true });

            const cmd = execa("node", ["../cli/cli", "generate", fixturePath], {
                env: {
                    NODE_ENV: "development",
                },
            });
            cmd.stdout?.pipe(process.stdout);
            cmd.stderr?.pipe(process.stderr);
            await cmd;

            await installAndCompileGeneratedProject(outputPath);

            const expectedFilesBuffer = await readFile(path.join(fixturePath, "expectedFiles.txt"));
            const expectedFiles = expectedFilesBuffer
                .toString()
                .split("\n")
                .map((s) => s.trim())
                .filter((s) => s.length > 0);

            for (const expectedFile of expectedFiles) {
                let fileContents: string;
                try {
                    fileContents = (await readFile(path.join(outputPath, expectedFile))).toString();
                } catch (e) {
                    throw new Error(`Expected file ${expectedFile} to exist, but it does not.`);
                }
                expect(fileContents).toMatchSnapshot();
            }
            await rm(outputPath, { force: true, recursive: true });
        },
        90_000
    );
}
