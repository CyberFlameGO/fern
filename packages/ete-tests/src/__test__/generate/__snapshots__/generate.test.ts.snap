// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fern generate tests simple-model 1`] = `Array []`;
